src/Illuminate/Database/Eloquent/Relations/HasOne.php

```diff
@@ -77,12 +76,22 @@ public function match(array $models, Collection $results, $relation)
      */
     protected function getDefaultFor(Model $model)
     {
+        $result = null;
+
         if (is_callable($this->withDefault)) {
-            return call_user_func($this->withDefault);
+            $result = call_user_func($this->withDefault);
```

Comment:
@taylorotwell done. If nothing is returned from the closure, the default instance will be returned automatically.

Thank you @JosephSilber! The code looks better now.